### DISCUSSION TITLE
Passing varchar() parameters for String() columns in mssql

### DISCUSSION THREAD

**Discussion by gordthompson** [sqlalchemy/sqlalchemy#8171] [Usage Questions]
A [question posted on Gitter](https://gitter.im/sqlalchemy/community?at=62b4d17c1227f62be3b42959) relates to parameter values being passed by `mssql+pyodbc://`. In its simplest form,

```python
# fmt: off
import pyodbc
from sqlalchemy import create_engine, Table, MetaData, Column, String, event, update, Unicode
# fmt: on

engine = create_engine(
    "mssql+pyodbc://scott:tiger^5HHH@mssql_199",
)

my_table = Table(
    "my_table",
    MetaData(),
    Column("id", String(16), primary_key=True),
    Column("something", Unicode(10)),
)
my_table.drop(engine, checkfirst=True)
my_table.create(engine)

with engine.begin() as conn:
    upd = (
        update(my_table).where(my_table.c.id == "foo").values(something="bar")
    )
    conn.execute(upd)
```

both parameters are passed as Unicode, despite the fact that one is for a String() column and the other is for a Unicode() column.

```
declare @p1 int
set @p1=2
exec sp_prepexec @p1 output,N'@P1 nvarchar(6),@P2 nvarchar(6)',N'UPDATE my_table SET something=@P1 WHERE my_table.id = @P2',N'bar',N'foo'
select @p1
```

Note that both parameters are declared as `nvarchar(6)` and the values are passed as N'strings'.

This is a problem when the `id` column really is varchar(16) and not nvarchar(16) because the Unicode parameter value requires an index scan which degrades perfomance.

```
Clustered Index Scan (Clustered)
[my_table].[PK__my_table__3213E83FAE8EF454]
```

If the parameter value is passed as varchar then SQL Server can do a Seek instead of a Scan, which is significantly faster.

```
Clustered Index Seek (Clustered)
[my_table].[PK__my_table__3213E83FAE8EF454]
```

The fix in plain pyodbc is to use `setinputsizes()` to make the parameter `pyodbc.SQL_VARCHAR` instead of `pyodbc.SQL_WVARCHAR`. (`pyodbc.SQL_WVARCHAR` is the default because in Python_3 all strings are Unicode.)

In SQLAlchemy, using `use_setinputsizes=True` and dumping the `inputsizes` shows that the BindParameter entries are different:

```python
@event.listens_for(engine, "do_setinputsizes")
def receive_do_setinputsizes(
    inputsizes, cursor, statement, parameters, context
):
    print([repr(bind_param.type) for bind_param in inputsizes])
    # ['Unicode(length=10)', 'String(length=16)']
```

and one way to fix the issue is to modify the dict values for String() BindParameter entries

```python
@event.listens_for(engine, "do_setinputsizes")
def receive_do_setinputsizes(
    inputsizes, cursor, statement, parameters, context
):
    for bind_param in [
        x for x in inputsizes if repr(x.type).startswith("String(")
    ]:
        inputsizes[bind_param] = pyodbc.SQL_VARCHAR
```

resulting in

```
declare @p1 int
set @p1=2
exec sp_prepexec @p1 output,N'@P1 nvarchar(6),@P2 varchar(6)',N'UPDATE my_table SET something=@P1 WHERE my_table.id = @P2',N'bar','foo'
select @p1
```

but is this something that the dialect should be doing on its own? (That is, respecting the difference between String() columns and Unicode() columns when the BindParameter entries are processed.)

**Comment by zzzeek** [comment]
> The fix in plain pyodbc is to use `setinputsizes()` to make the parameter `pyodbc.SQL_VARCHAR` instead of `pyodbc.SQL_WVARCHAR`. (`pyodbc.SQL_WVARCHAR` is the default because in Python_3 all strings are Unicode.)

can you also just pass the string as bytes?   e.g. ``table.c.id == "foo".encode('ascii')``.   

> but is this something that the dialect should be doing on its own? (That is, respecting the difference between String() columns and Unicode() columns when the BindParameter entries are processed.)

try working with this patch which we can do for 2.0.  the weird thing about it is that it wont be enabled unless setinputsizes=True.   

```diff
diff --git a/lib/sqlalchemy/dialects/mssql/pyodbc.py b/lib/sqlalchemy/dialects/mssql/pyodbc.py
index 6d64fdc3ed..68ebd0e9c5 100644
--- a/lib/sqlalchemy/dialects/mssql/pyodbc.py
+++ b/lib/sqlalchemy/dialects/mssql/pyodbc.py
@@ -313,6 +313,8 @@ import re
 import struct
 
 from .base import _MSDateTime
+from .base import _MSUnicode
+from .base import _MSUnicodeText
 from .base import BINARY
 from .base import DATETIMEOFFSET
 from .base import MSDialect
@@ -465,6 +467,18 @@ class _VARBINARY_pyodbc(_ms_binary_pyodbc, VARBINARY):
 class _BINARY_pyodbc(_ms_binary_pyodbc, BINARY):
     pass
 
+class _String_pyodbc(sqltypes.String):
+    def get_dbapi_type(self, dbapi):
+        return dbapi.SQL_VARCHAR
+
+class _Unicode_pyodbc(_MSUnicode):
+    def get_dbapi_type(self, dbapi):
+        return dbapi.SQL_WVARCHAR
+
+class _UnicodeText_pyodbc(_MSUnicodeText):
+    def get_dbapi_type(self, dbapi):
+        return dbapi.SQL_WVARCHAR
+
 
 class MSExecutionContext_pyodbc(MSExecutionContext):
     _embedded_scope_identity = False
@@ -541,6 +555,10 @@ class MSDialect_pyodbc(PyODBCConnector, MSDialect):
             VARBINARY: _VARBINARY_pyodbc,
             sqltypes.VARBINARY: _VARBINARY_pyodbc,
             sqltypes.LargeBinary: _VARBINARY_pyodbc,
+
+            sqltypes.String: _String_pyodbc,
+            sqltypes.Unicode: _Unicode_pyodbc,
+            sqltypes.UnicodeText: _UnicodeText_pyodbc,
         },
     )
 
```

**Reply by gordthompson** [reply]
Yup, your patch achieves the same result as my event thingy. And yes, the patch currently requires `create_engine(connection_url, use_setinputsizes=True)` to take effect.

However, I also notice that the MCVE code from

https://github.com/sqlalchemy/sqlalchemy/issues/6058#issuecomment-800284593

does not fail under the current (unpatched) main branch (6f3a798) or under 1.4.39. The parameters are correctly passed as `numeric(9,6)`

```
declare @p1 int
set @p1=3
exec sp_prepare @p1 output,N'@P1 int,@P2 numeric(9,6),@P3 numeric(9,6)',N'INSERT INTO test_schema.[Observations] (id, obs1, obs2) VALUES (@P1, @P2, @P3)',1
select @p1
```

so maybe we can consider letting `mssql+pyodbc://` default to `use_setinputsizes=True` again.

**Reply by zzzeek** [reply]
I would love to just have setinputsizes turned on and that's it.  

lets make a new issue and a gerrit for 2.0 but also we need to understand why #6058 stopped being a problem (change in pyodbc?)

**Reply by gordthompson** [reply]
Sorry, sloppy testing on my part. The MCVE still fails under 1.4.39 if we remember to `use_setinputsizes=True`. 🤦‍♂️

**Reply by zzzeek** [reply]
and what about 2.0?

**Reply by zzzeek** [reply]
do we want to just have setinputsizes True unless fast_executemany is turned on ?

**Reply by gordthompson** [reply]
I think it would be reasonable to have `use_setinputsizes=True` as the default in 2.0 but if they specify `fast_executemany=True` then force `use_setinputsizes=False` (and perhaps emit a warning). setinputsizes doesn't really apply in the typical usage case for fast_executemany because fast_executemany causes pyodbc to retrieve the column types from the server.

I suppose if people really need both features in the same app they could use two engines:

```python
engine = create_engine(connection_url)  # use_setinputsizes=True by default
fast_executemany_engine = create_engine(connection_url, fast_executemany=True)  # forces use_setinputsizes=False
```

**Reply by zzzeek** [reply]
do we want to just have setinputsizes skipped explicitly for cursor.executemany() when fast_executemany is turned on instead?  fast_executemany does not affect cursor.execute(), right?

**Reply by gordthompson** [reply]
> do we want to just have setinputsizes skipped explicitly for cursor.executemany() when fast_executemany is turned on instead?

Hey, that's a good idea. fast_executemany is usually used for inserts but if they used it on an update (as in the original post) then they'd be doing index scans but at least SQL Server wouldn't barf.

Besides, for newer versions of SQL Server there's an alternative to fast_executemany

https://github.com/mkleehammer/pyodbc/wiki/Tips-and-Tricks-by-Database-Platform#passing-row-oriented-parameter-data-as-a-json-string

> fast_executemany does not affect cursor.execute(), right?

Right.

**Comment by gordthompson** [comment]
@amygurski - Here's a workaround for your specific issue while we get this sorted out.

You are using a `TextClause`

```python
sql = text('UPDATE My_Table SET something = :bar WHERE id = :foo')
```

and AFAIK the current SQLAlchemy implementation of setinputsizes doesn't apply to those anyway (right, @zzzeek ?) but you could use

```python
sql = text('UPDATE My_Table SET something = :bar WHERE id = CAST(:foo AS varchar(16))')
conn.execute(sql, dict(bar="bar_value", foo="foo_value"))
```

instead.

**Reply by zzzeek** [reply]
@gordthompson can you create an issue and milestone it to 2.0 "deafult setinputsizes=True for SQL Server pyodbc and automate it against fast_executemany"?  thanks

**Reply by gordthompson** [reply]
Okay, done. https://github.com/sqlalchemy/sqlalchemy/issues/8177

**Reply by amygurski** [reply]
Thank you so much for reponses and help!

We were able to re-write one app with minimal effort to be `TextClause` so we could just cast the strings as varchar. 75% of the problematic queries were already textual sql anyway so this worked swimmingly and we were able to upgrade from python 3.6 -> 3.8 which is where this python string -> mssql nvarchar issue crops up.

We are still struggling to get our other app working on python 3.8. This app uses the ORM setup with `query` objects and a `scoped_session`. We are trying to implement the `do_setinputsizes` event to capture the strings and ensure the dbapi type is varchar.

When we put the event listener right after `create_engine` (where it feels like it should go) then it doesn't work - it never even hits the listener:
```
engine = create_engine(..., use_setinputsizes=True)

@event.listens_for(engine, 'do_setinputsizes')
def receive_do_setinputsizes(inputsizes, cursor, statement, parameters, context):
    # logging here is never reached; nvarchar arriving at mssql
    for bind_param in [ bp for bp in inputsizes if repr(bp.type).startswith("String(") ]:
        inputsizes[bind_param] = pyodbc.SQL_VARCHAR

db_session = scoped_session(
    sessionmaker(
        autoflush=True,
        autocommit=False,
    ))

db_session.configure(bind=engine)
```
When we move the event listener to the `BaseSQL` model that models inherit from it works beautifully:
```
BaseDec = declarative_base()

class BaseSQL(object):
    @classmethod
    def query(cls):
       session = cls.session()

       @event.listens_for(session.get_bind(), 'do_setinputsizes')
       def receive_do_setinputsizes(inputsizes, cursor, statement, parameters, context):
           for bind_param in [ bp for bp in inputsizes if repr(bp.type).startswith("String(") ]:
               # works beautifully - no nvarchar in sight in mssql, even where cls.query() isn't called, i.e. the relationships.
               inputsizes[bind_param] = pyodbc.SQL_VARCHAR

        return session.query(cls)

class User(BaseDec, BaseSQL):
    __tablename__ = 'User'
    
    name = Column(String(16))
    requests = relationship('Request', backref = 'something')
  
    @classmethod
    def find(cls):
        return cls.query().filter(cls.name==name).first()
```
This has a 100% success rate in that the pesky nvarchars are eradicated. However, we occasionally hit a `deque mutated during iteration` error when we commit the transactions which causes our app to return a 500. From what I've read this is intentional in sqlalchemy and means the collection is being mutated while it's being iterated.

If you have any additional pointers for what we may be doing wrong or info we could provide to help troubleshoot it would be really appreciated!

**Reply by zzzeek** [reply]
firstly it seems like session.get_bind() is returning a different Engine object than the one you are showing with the create_engine(), so you'd want to check on that.   

re: "deque mutated during iteration" , this is probably because you are attaching the event listener on the fly within Query - that's going to ultimately cause you a memory leak since you are re-attaching the same event listener over and over again, as well as degrade performance very quickly.   you need to figure out what "engine" is bound to your Session in the first place, and attach the event there.

Alternatively, you can associate the event with the `sqlalchemy.engine.Engine` class overall.

**Reply by amygurski** [reply]
You called it! The engine we'd tried to associate the event listener with directly after `create_engine` was called was only used by a background worker. The web requests had a custom cherrypy tool (located in a separate internal library) enabled in the configs. It was inside that tool that the engine we cared about attaching the event listener to was getting instantiated 🤦‍♀️ . So we are up and running! 🎉 Thanks again!

Is it possible to write a unittest confirming that the dbapi type is now VARCHAR instead of NVARCHAR? When we log inside the event listener, we see that the `inputsizes[bind_param]` changes from `class <str>` to `12` (`pyodbc.SQL_VARCHAR`), but we haven't had any luck finding/accessing `inputsizes` outside the event listener for example.

**Reply by zzzeek** [reply]
well if pyodbc.cursor doesnt give you a way to read that, you'd have to dig in and make a Mock() of the cursor and send it through.    alternatively, perhaps there's a query you could run, which will receive the setinputsizes command you want to test, and then somehow there's something about the statement itself that can reveal how the data was bound.  sounds a little weird,but overall you need to figure out how you'd idealy detect "success".
